File: src/python/xerxes/streaming/sse.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Incremental Server-Sent-Events parser with ``Last-Event-ID`` support.

Used by the streaming-provider clients to recover from transient drops in the
middle of a Responses-API stream. The parser is pure — it consumes raw text
chunks via :meth:`SSEParser.feed`, releases completed records via
:meth:`SSEParser.drain`, and tracks ``last_event_id`` for use as the
``Last-Event-ID`` HTTP header on reconnect. The caller decides whether to
reconnect or abort.
"""

from __future__ import annotations

from collections.abc import Iterable, Iterator
from dataclasses import dataclass


@dataclass
class SSEEvent:
    """A single completed SSE record.

    Attributes:
        event: Event name (the value of the ``event:`` field; defaults to
            ``"message"`` per the spec).
        data: Concatenated value of all ``data:`` lines in the record.
        id: Event id (the ``id:`` field), used for ``Last-Event-ID`` resume.
        retry: Suggested reconnect delay in milliseconds, or ``None``.
    """

    event: str = "message"
    data: str = ""
    id: str = ""
    retry: int | None = None


class SSEParser:
    """Incremental Server-Sent-Events parser.

    Feed raw text chunks via :meth:`feed`; pop completed records with
    :meth:`drain`. The parser remembers the last ``id`` it saw on
    :attr:`last_event_id` so the caller can pass it as ``Last-Event-ID`` when
    reconnecting after a drop.
    """

    def __init__(self) -> None:
        """Initialise empty buffers for a fresh stream."""
        self._buffer = ""
        self._current_event: str = "message"
        self._current_data: list[str] = []
        self._current_id: str = ""
        self._current_retry: int | None = None
        self._completed: list[SSEEvent] = []
        self.last_event_id: str = ""

    def feed(self, chunk: str) -> None:
        """Buffer a chunk of raw text and process every complete line."""
        self._buffer += chunk
        while "\n" in self._buffer:
            line, _, rest = self._buffer.partition("\n")
            self._buffer = rest
            self._handle_line(line.rstrip("\r"))

    def _handle_line(self, line: str) -> None:
        """Apply one parsed SSE line; blank lines flush the current record."""
        if not line:
            self._dispatch()
            return
        if line.startswith(":"):
            return  # comment
        if ":" in line:
            field_name, _, value = line.partition(":")
            value = value.lstrip(" ")
        else:
            field_name, value = line, ""
        if field_name == "event":
            self._current_event = value
        elif field_name == "data":
            self._current_data.append(value)
        elif field_name == "id":
            self._current_id = value
        elif field_name == "retry":
            try:
                self._current_retry = int(value)
            except ValueError:
                self._current_retry = None

    def _dispatch(self) -> None:
        """Finalise the in-progress record into an :class:`SSEEvent`."""
        if not self._current_data and not self._current_event != "message":
            return
        ev = SSEEvent(
            event=self._current_event,
            data="\n".join(self._current_data),
            id=self._current_id,
            retry=self._current_retry,
        )
        self._completed.append(ev)
        if ev.id:
            self.last_event_id = ev.id
        self._current_event = "message"
        self._current_data = []
        self._current_id = ""
        self._current_retry = None

    def drain(self) -> list[SSEEvent]:
        """Return and clear all events completed since the last drain."""
        out = self._completed
        self._completed = []
        return out


def parse_sse_stream(chunks: Iterable[str]) -> Iterator[SSEEvent]:
    """Parse a finite chunk stream end-to-end and yield every event.

    Convenience over :class:`SSEParser` for tests and one-shot uses. After
    feeding the chunks, a trailing ``\\n\\n`` is appended to flush any final
    record that lacks a terminating blank line.
    """
    p = SSEParser()
    for chunk in chunks:
        p.feed(chunk)
        for ev in p.drain():
            yield ev
    # Flush any pending event (final lines without trailing blank line).
    p.feed("\n\n")
    for ev in p.drain():
        yield ev


__all__ = ["SSEEvent", "SSEParser", "parse_sse_stream"]

File: xerxes/src/streaming/sse.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

/** A completed Server-Sent Events record. */
export interface SSEEvent {
  readonly data: string
  readonly event: string
  readonly id: string
  readonly retry: number | undefined
}

/**
 * Incrementally parse a Server-Sent Events text stream.
 *
 * Feed decoded text chunks as they arrive, then retrieve complete records with
 * {@link drain}. `lastEventId` is available to callers that reconnect with a
 * `Last-Event-ID` request header.
 */
export class SSEParser {
  private buffer = ''
  private completed: SSEEvent[] = []
  private currentData: string[] = []
  private currentEvent = 'message'
  private currentId = ''
  private currentRetry: number | undefined

  lastEventId = ''

  /** Buffer raw text and process every complete line. */
  feed(chunk: string): void {
    this.buffer += chunk
    for (;;) {
      const newline = this.buffer.indexOf('\n')
      const carriage = this.buffer.indexOf('\r')
      let end = -1
      let skip = 1
      if (newline >= 0 && (carriage < 0 || newline < carriage)) {
        end = newline
      } else if (carriage >= 0) {
        // A bare '\r' terminates a line, but a trailing '\r' may be half of a
        // '\r\n' pair split across chunks, so wait for the next chunk.
        if (carriage === this.buffer.length - 1) break
        end = carriage
        skip = this.buffer[carriage + 1] === '\n' ? 2 : 1
      } else {
        break
      }
      const line = this.buffer.slice(0, end)
      this.buffer = this.buffer.slice(end + skip)
      this.handleLine(line)
    }
  }

  /** Return and clear the records completed since the preceding drain. */
  drain(): SSEEvent[] {
    const completed = this.completed
    this.completed = []
    return completed
  }

  private handleLine(line: string): void {
    if (!line) {
      this.dispatch()
      return
    }
    if (line.startsWith(':')) {
      return
    }

    const separator = line.indexOf(':')
    const field = separator < 0 ? line : line.slice(0, separator)
    const rawValue = separator < 0 ? '' : line.slice(separator + 1)
    // The SSE spec removes exactly one leading space after the colon.
    const value = rawValue.startsWith(' ') ? rawValue.slice(1) : rawValue
    if (field === 'event') {
      this.currentEvent = value
    } else if (field === 'data') {
      this.currentData.push(value)
    } else if (field === 'id') {
      this.currentId = value
    } else if (field === 'retry') {
      this.currentRetry = parseRetry(value)
    }
  }

  private dispatch(): void {
    // Per the WHATWG SSE spec, an event is never dispatched with an empty
    // data buffer, regardless of the event type.
    if (!this.currentData.length) {
      this.resetCurrentRecord()
      return
    }

    const event: SSEEvent = {
      event: this.currentEvent,
      data: this.currentData.join('\n'),
      id: this.currentId,
      retry: this.currentRetry,
    }
    this.completed.push(event)
    if (event.id) {
      this.lastEventId = event.id
    }
    this.resetCurrentRecord()
  }

  private resetCurrentRecord(): void {
    this.currentData = []
    this.currentEvent = 'message'
    this.currentId = ''
    this.currentRetry = undefined
  }
}

/**
 * Parse a finite sequence of decoded SSE chunks.
 *
 * A final blank record delimiter flushes an otherwise unterminated final
 * event, matching the incremental parser's normal record boundary behavior.
 */
export function* parseSseStream(chunks: Iterable<string>): Generator<SSEEvent> {
  const parser = new SSEParser()
  for (const chunk of chunks) {
    parser.feed(chunk)
    yield* parser.drain()
  }
  parser.feed('\n\n')
  yield* parser.drain()
}

function parseRetry(value: string): number | undefined {
  const normalized = value.trim()
  if (!/^[+-]?\d+$/.test(normalized)) {
    return undefined
  }
  const retry = Number(normalized)
  return Number.isSafeInteger(retry) ? retry : undefined
}
